xdsl/backend/csl/lowering/convert_stencil_to_csl.py

```diff
@@ -0,0 +1,399 @@
+                "computeAllRoutes",
+                ComptimeStructType(),
+                routes,
+                x_id,
+                y_id,
+                width,
+                height,
+                pattern_op,
+            ),
+            # setting up param structs for `@set_tile_code`
+            set_tile_params := ConstStructOp.get(
+                (Named.memcpy_params, memcpy_params),
+                (Named.stencil_comms_params, route_params),
+            ),
+            params_task := ConstStructOp.get((Named.is_border_region_pe, is_border_pe)),
+            set_tile_params_ext := ConcatStructOp.get(params_task, set_tile_params),
+            # generates: @set_tile_code(xId, yId, "pe.csl", .. param structs ..)
+            SetTileCodeOp.get(
+                ctx.program_module.sym_name,
+                x_id,
+                y_id,
+                set_tile_params_ext,
+            ),
+            Yield(),
+        ]
+    )
+    outer_loop_body.add_op(Yield())
+
+
+@dataclass(frozen=True)
+class ConvertStencilToCsl(ModulePass):
+    name = "convert-stencil-to-csl"
+
+    def apply(self, ctx: MLContext, op: ModuleOp) -> None:
+        # initialise both CSL modules
+        t_ctx = TranslationContext()
+        t_ctx.program_module = CslModuleOp(
+            regions=[Region(Block())],
+            properties={"kind": ModuleKindAttr(ModuleKind.PROGRAM)},
+            attributes={"sym_name": StringAttr("pe.csl")},
+        )
+        t_ctx.layout_module = CslModuleOp(
+            regions=[Region(Block())],
+            properties={"kind": ModuleKindAttr(ModuleKind.LAYOUT)},
+            attributes={"sym_name": StringAttr("layout.csl")},
+        )
+        # iterate over stencil.apply ops to find the maximum stencil width ('arm length')
+        assert len(op.body.block.ops) == 1
+        assert isinstance(stencil_func := op.body.block.first_op, func.FuncOp)
+        t_ctx.program_sym_name = stencil_func.sym_name
+        neighbours = 0
+        for o in stencil_func.body.block.ops:
+            if isinstance(o, ApplyOp):
+                for access in o.get_accesses():
+                    for l, r in access.halos():
+                        neighbours = max(neighbours, abs(l), abs(r))
+        # the CSL `pattern` variable corresponds to the neighbours in each direction plus 1
+        t_ctx.pattern = neighbours + 1
+
+        # determine the grid dimensions by looking at the function args
+        for arg in stencil_func.args:
+            if isa(arg.type, MemRefType[Attribute]) and isa(
+                arg.type.element_type, TensorType[Attribute]
+            ):
+                shape = arg.type.get_shape()
+                t_ctx.grid_dim = tuple(
+                    max(a, b)
+                    for a, b in zip(
+                        t_ctx.grid_dim,
+                        (shape[0], shape[1], arg.type.element_type.get_shape()[0]),
+                    )
+                )
```

Comment:
Let's think about how we actually want this conversion to happen. I don't think this is a very scalable approach.

Reply:
I'm afraid we will need some variant of this, as the communications library is bound to a fixed stencil size (which I suppose needs to be the maximum we encounter throughout the program).

The grid dimensions are primarily (though not exclusively) needed for the 2-dimensional loop performed in layout.csl to `set_tile_code` for each PE. The idea here is to derive the size of the compute grid from the size of the buffers passed in to the stencil function - the maximum size, to be safe.